DatabaseServer/database_server.py

```diff
@@ -341,22 +350,21 @@ def _get_iocs_not_to_stop() -> list:
 
     # Initialise IOC database connection
     try:
-        ioc_data = IOCData(IocDataSource(SQLAbstraction("iocdb", "iocdb", "$iocdb")), ProcServWrapper(),
-                           MACROS["$(MYPVPREFIX)"])
+        ioc_data = init_ioc_database_connection()
         print_and_log("Connected to IOCData database", INFO_MSG, LOG_TARGET)
-    except Exception as e:
+    except MaxAttemptsExceededException as ex:
         ioc_data = None
-        print_and_log("Problem initialising IOCData DB connection: {}".format(traceback.format_exc()),
-                      MAJOR_MSG, LOG_TARGET)
+        exception_info = "".join(traceback.format_exception(type(ex.args[0]), ex.args[0], ex.args[0].__traceback__))
+        print_and_log(f"Problem initialising IOCData DB connection: {exception_info}", MAJOR_MSG, LOG_TARGET)
 
     # Initialise experimental database connection
     try:
-        exp_data = ExpData(MACROS["$(MYPVPREFIX)"], ExpDataSource())
+        exp_data = init_experimental_database_connection()
         print_and_log("Connected to experimental details database", INFO_MSG, LOG_TARGET)
-    except Exception as e:
+    except MaxAttemptsExceededException as ex:
         exp_data = None
-        print_and_log("Problem connecting to experimental details database: {}".format(traceback.format_exc()),
-                      MAJOR_MSG, LOG_TARGET)
+        exception_info = "".join(traceback.format_exception(type(ex.args[0]), ex.args[0], ex.args[0].__traceback__))
+        print_and_log(f"Problem connecting to experimental details database: {exception_info}", MAJOR_MSG, LOG_TARGET)
 
     DRIVER = DatabaseServer(SERVER, ioc_data, exp_data, OPTIONS_DIR, BLOCKSERVER_PREFIX)
```

Comment:
Both `ioc_data` and `exp_data` could be `None` here. I think we should _explicitly_ handle this case, or prevent it from happening.